lib/qless/middleware/timeout.rb

```diff
@@ -15,29 +16,45 @@ module Middleware
     # or a variable timeout based on the individual TTLs of each job
     # (using something like `extend Qless::Middleware::Timeout.new { |job| job.ttl * 1.1 }`).
     class Timeout < Module
-      def initialize
+      def initialize(opts = {})
+        timeout_class = opts.fetch(:timeout_class, ::Timeout)
+        kernel_class = opts.fetch(:kernel_class, Kernel)
         module_eval do
           define_method :around_perform do |job|
-            timeout_value = yield job
-            if timeout_value <= 0
-              raise InvalidTimeoutError, "Timeout must be positive, but was #{timeout_value}"
-            end
+            timeout_seconds = yield job
+            if timeout_seconds.nil?
+              super(job)
+            else
+              if !timeout_seconds.is_a?(Numeric) || timeout_seconds <= 0
+                raise InvalidTimeoutError, "Timeout must be a positive number or nil, " \
+                                           "but was #{timeout_seconds}"
+              end
 
-            begin
-              ::Timeout.timeout(timeout_value) { super(job) }
-            rescue ::Timeout::Error => e
-              error = JobTimedoutError.new(e.message)
-              error.set_backtrace(e.backtrace)
-              job.fail(*Qless.failure_formatter.format(job, error, []))
+              begin
+                timeout_class.timeout(timeout_seconds) { super(job) }
+              rescue ::Timeout::Error => e
+                error = JobTimedoutError.new(e.message)
+                error.set_backtrace(e.backtrace)
+                # The stalled connection to redis might be the cause of the timeout. We cannot rely
+                # on state of connection either (e.g., we might be in the middle of Redis call when
+                # timeout happend). To play it safe, we reconnect.
+                job.reconnect_to_redis
+                job.fail(*Qless.failure_formatter.format(job, error, []))
+                # Since we are leaving with bang (exit!), normal requeue logic does not work.
+                # Do it manually right here.
+                if self.is_a?(::Qless::Middleware::RequeueExceptions) && self.requeueable?(JobTimedoutError)
```

Comment:
Here is my design constrains:
* the both middlewares are in the same package (`Qless`), so they are delivered together.
* the timeout is very special case in the sense that it occurs outside of normal execution context. In particular it might have socket in weird state (including redis one). So we need to harden timeout processing logic for weird situations. In the same time, normal requeue does not need it.
* The (causing timeout) issue can be with Qless (redis) itself (and we saw it). So we need to protect against it. Part of it is `reconnect`. The Qless portion should be protected by secondary timeout, but I would rather implement is as a separate PR to avoid confusion.
* We want to avoid dependency of ordering requeue and timeout middleware.
* Thus some of the timeout logic must happen before, and some after requeue logic. 
* We want very loose coupling of two middlewares. So, while timeout middleware delegates to requeue one, the worker might be configured to be in any of four combinations.